base/compiler/abstractinterpretation.jl

```diff
@@ -571,17 +605,80 @@ struct MethodCallResult
     edgecycle::Bool
     edgelimited::Bool
     edge::Union{Nothing,MethodInstance}
+    edge_effects::Effects
     function MethodCallResult(@nospecialize(rt),
                               edgecycle::Bool,
                               edgelimited::Bool,
-                              edge::Union{Nothing,MethodInstance})
-        return new(rt, edgecycle, edgelimited, edge)
+                              edge::Union{Nothing,MethodInstance},
+                              edge_effects::Effects)
+        return new(rt, edgecycle, edgelimited, edge, edge_effects)
+    end
+end
+
+function is_all_const_arg((; fargs, argtypes)::ArgInfo)
+    for a in argtypes
+        if !isa(a, Const) && !isconstType(a) && !issingletontype(a)
+            return false
+        end
     end
+    return true
+end
+
+function concrete_eval_const_proven_total_or_error(
+        interp::AbstractInterpreter,
+        @nospecialize(f), argtypes::Vector{Any})
+    args = Any[ (a = widenconditional(argtypes[i]);
+        isa(a, Const) ? a.val :
+        isconstType(a) ? (a::DataType).parameters[1] :
+                         (a::DataType).instance) for i in 2:length(argtypes) ]
+    try
+        value = Core._call_in_world_nonpure(get_world_counter(interp), f, args...)
+        return Const(value)
+    catch e
+        return nothing
+    end
+end
+
+function const_prop_enabled(interp::AbstractInterpreter, sv::InferenceState, match::MethodMatch)
+    if !InferenceParams(interp).ipo_constant_propagation
+        add_remark!(interp, sv, "[constprop] Disabled by parameter")
+        return false
+    end
+    method = match.method
+    if method.constprop == 0x02
+        add_remark!(interp, sv, "[constprop] Disabled by method parameter")
+        return false
+    end
+    return true
+end
+
+struct ConstCallResults
+    rt::Any
+    const_result::Union{InferenceResult, ConstResult}
+    effects::Effects
+    ConstCallResults(@nospecialize(rt),
+                     const_result::Union{InferenceResult, ConstResult},
+                     effects::Effects) =
+        new(rt, const_result, effects)
 end
 
 function abstract_call_method_with_const_args(interp::AbstractInterpreter, result::MethodCallResult,
                                               @nospecialize(f), arginfo::ArgInfo, match::MethodMatch,
                                               sv::InferenceState, va_override::Bool)
+    if !const_prop_enabled(interp, sv, match)
+        return nothing
+    end
+    if f !== nothing && result.edge !== nothing && is_total_or_error(result.edge_effects) && is_all_const_arg(arginfo)
+        rt = concrete_eval_const_proven_total_or_error(interp, f, arginfo.argtypes)
+        add_backedge!(result.edge, sv)
+        rt === nothing && return ConstCallResults(Union{}, ConstResult(result.edge), result.edge_effects) # The evaulation threw. By :consistent-cy, we're guaranteed this would have happened at runtime
```

Comment:
```suggestion
        if rt === nothing
            # The evaluation threw. By :consistent-cy, we are guaranteed this would have happened at runtime.
            return ConstCallResults(Union{}, ConstResult(result.edge), result.edge_effects)
        end
```